spec-tests/src/test/java/org/candlepin/spec/PoolSpecTest.java

```diff
@@ -0,0 +1,673 @@
+import org.candlepin.spec.bootstrap.data.builder.Owners;
+import org.candlepin.spec.bootstrap.data.util.UserUtil;
+
+import org.junit.jupiter.api.Test;
+
+import java.time.OffsetDateTime;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+
+
+
+@SpecTest
+class PoolSpecTest {
+
+    /**
+     * Generates a random suffix of 7 characters or fewer. Alternative to StringUtil.random which
+     * generates too many characters for fields which have small length limitations.
+     *
+     * @param prefix
+     *  The base string to concatenate to the generated suffix
+     *
+     * @return
+     *  the provided base string/prefix concatenated with a randomly generated suffix
+     */
+    private String randomSuffix(String prefix) {
+        return prefix + "-" + ((int) (Math.random() * 100000));
+    }
+
+    private ApiClient createUserClient(ApiClient client, OwnerDTO owner) throws ApiException {
+        UserDTO user = UserUtil.createUser(client, owner);
+        return ApiClients.trustedUser(user.getUsername(), true);
+    }
+
+    private ApiClient createConsumerClient(ApiClient client, OwnerDTO owner) throws ApiException {
+        String consumerName = this.randomSuffix("test_consumer");
+        ConsumerDTO consumer = this.createConsumer(client, owner, consumerName, "system", null);
+
+        return this.createConsumerClient(client, consumer);
+    }
+
+    private ApiClient createConsumerClient(ApiClient client, ConsumerDTO consumer) throws ApiException {
+        return ApiClients.ssl(consumer.getIdCert());
+    }
+
+    private OwnerDTO createOwner(ApiClient client) throws ApiException {
+        return client.owners().createOwner(Owners.random());
+    }
+
+    private ConsumerDTO createConsumer(ApiClient client, OwnerDTO owner, String name, String type,
+        Map<String, String> facts) throws ApiException {
+
+        ConsumerTypeDTO ctype = new ConsumerTypeDTO()
+            .label(type);
+
+        ConsumerDTO consumer = new ConsumerDTO()
+            .name(name)
+            .owner(Owners.toNested(owner))
+            .type(ctype);
+
+        if (facts != null) {
+            // Copy the map into a new hashmap to ensure we can still modify it later if we so choose
+            consumer.facts(new HashMap<>(facts));
+        }
+
+        consumer.putFactsItem("system.certificate_version", "3.3");
+
+        return client.consumers()
+            .register(consumer);
+    }
+
+    private ProductDTO createProduct(ApiClient client, OwnerDTO owner) throws ApiException {
+        String productId = String.valueOf((int) Math.random() * 10000);
+
+        ProductDTO product = new ProductDTO()
+            .id(productId)
+            .name("test_product-" + productId);
+
+        return client.ownerProducts()
+            .createProductByOwner(owner.getKey(), product);
+    }
+
+    // ugh...
+    private AttributeDTO buildAttribute(String key, String value) {
```

Comment:
Check ProductAttributes

Reply:
ProductAttributes currently lives as an enum which seems like it's trying to programmatically standardize the attribute names (as we do in CP code), but the way you have to work with it to build attribute DTOs is still clunky.

This method exists because I wanted a quick one-liner that would take a key-value pair and turn it into the object without in-line method chaining. With the ProductAttributes version, I basically end up with the same thing I already have, but with an additional wrapper to unwrap:

```
private AttributeDTO buildAttribute(String key, String value) {
    return new ProductAttributes(key)
        .withValue(value)
        .value();
}
```

There's probably more we could do here to improve that factory.